rust/azure_iot_operations_connector/src/base_connector.rs

```diff
@@ -2,3 +2,118 @@
+    azure_device_registry_client: azure_device_registry::Client<SessionManagedClient>,
+    // state_store_client: Arc<state_store::Client<SessionManagedClient>>,
+    // schema_registry_client: schema_registry::Client<SessionManagedClient>,
+    // etc
+}
+
+/// Base Connector for Azure IoT Operations
+pub struct BaseConnector {
+    connector_context: ConnectorContext,
+    session: Session,
+}
+
+impl BaseConnector {
+    /// Creates a new [`BaseConnector`] and all required clients/etc needed to run connector operations.
+    /// On any failures, will log the error and then retry getting the connector configuration
+    /// with exponential backoff. This allows for new configuration to be deployed to fix any
+    /// errors without needing to restart the connector.
+    #[must_use]
+    pub fn new(application_context: &ApplicationContext) -> Self {
+        // if any of these operations fail, wait and try again in case connector configuration has changed
+        operation_with_retries::<Self, String>(|| {
+            // Get Connector Configuration
+            let connector_config =
+                ConnectorConfiguration::new_from_deployment().map_err(|e| e.to_string())?;
+
+            // Create Session
+            let mqtt_connection_settings = connector_config
+                .clone()
+                .to_mqtt_connection_settings("0")
+                .map_err(|e| e.to_string())?;
+            let session_options = SessionOptionsBuilder::default()
+                .connection_settings(mqtt_connection_settings.clone())
+                // TODO: reconnect policy
+                // TODO: outgoing_max
+                .build()
+                .map_err(|e| e.to_string())?;
+            let session = Session::new(session_options).map_err(|e| e.to_string())?;
+
+            // Create clients
+            // Create Azure Device Registry Client
+            let azure_device_registry_client = azure_device_registry::Client::new(
+                application_context.clone(),
+                session.create_managed_client(),
+                azure_device_registry::ClientOptions::default(),
+            )
+            .map_err(|e| e.to_string())?;
+
+            Ok(Self {
+                connector_context: ConnectorContext {
+                    debounce_duration: Duration::from_secs(5), // TODO: come from somewhere
+                    application_context: application_context.clone(),
+                    connector_config,
+                    azure_device_registry_client,
+                },
+                session,
+            })
+        })
+    }
+
+    /// Runs the MQTT Session that allows all Connector Operations to be performed
+    /// Returns if the session ends. If this happens, the base connector will need to be recreated
+    ///
+    /// # Errors
+    /// Returns a [`SessionError`] if the session encounters a fatal error and ends.
+    pub async fn run(self) -> Result<(), SessionError> {
```

Comment:
Was the eventual decision to keep the new/run split? Or was that elsewhere?

Reply:
because this is now going to be something that is used by instead of owning the protocol translator, then new and run have to be separate so you can call the get_managed_... (which is used to do the actual fn calls) in between the two calls (run will be consuming)